Tutorial: Create and publish packages does not work

I'm working through the tutorial set.

I started with _Install and use packages with CMake_ and transitioned to this tutorial _Create and publish packages_. This package creation tutorial is not working.

The first observation I had was that the `sha` returned by the first `vcpkg install vcpkg-sample-library --overlay-ports=...` step doesn't match the `sha` in the documentation.

Next, after updating the `sha` the second `vcpkg install vcpkg-sample-library --overlay-ports=...` doesn't compile anything on my system, but the example shows that something is compiled.

Then verifying with the `helloworld` project fails during the `cmake --preset=default` stage.

System information and details follow. It's long winded with all the steps to illusrate I did what the example says to do.

# System Information
* Windows 11 Pro N (23H2 / 22631.3007)
* VS2017 (15.9.59), VS2019 (16.11.33) and VS2022 (17.8.4) are **_all_** installed
* cmake 3.23.1
* git 2.42.0.windows.2

# Details

## Install and Use Packages

```
c:\Users\rocha\Desktop\VCPKG-TEST>git clone https://github.com/microsoft/vcpkg.git
git clone https://github.com/microsoft/vcpkg.git
Cloning into 'vcpkg'...
Updating files: 100% (11074/11074), done.
```

```
c:\Users\rocha\Desktop\VCPKG-TEST>cd vcpkg
cd vcpkg

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>bootstrap-vcpkg.bat 
bootstrap-vcpkg.bat 
Downloading https://github.com/microsoft/vcpkg-tool/releases/download/2024-01-11/vcpkg.exe -> C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\vcpkg.exe... done.
Validating signature... done.

vcpkg package management program version 2024-01-11-710a3116bbd615864eef5f9010af178034cb9b44

See LICENSE.txt for license information.
Telemetry
---------
vcpkg collects usage data in order to help us improve your experience.
The data collected by Microsoft is anonymous.
You can opt-out of telemetry by re-running the bootstrap-vcpkg script with -disableMetrics,
passing --disable-metrics to vcpkg on the command line,
or by setting the VCPKG_DISABLE_METRICS environment variable.

Read more about vcpkg telemetry at docs/about/privacy.md

```

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>set "VCPKG_ROOT=%CD%"
set "VCPKG_ROOT=%CD%"

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>set VCPKG_ROOT
set VCPKG_ROOT
VCPKG_ROOT=c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>set "PATH=%PATH%;%VCPKG_ROOT%"
set "PATH=%PATH%;%VCPKG_ROOT%"

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>set PATH
set PATH
PATH=C:\Program Files\Eclipse Adoptium\jdk-11.0.21.9-hotspot\bin;C:\Program Files\Python310\Scripts\;C:\Program Files\Python310\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\Perforce\;C:\Program Files\IcedTeaWeb\WebStart\bin;C:\Program Files\CMake\bin;C:\Program Files (x86)\Graphviz2.38\bin;C:\Program Files\doxygen\bin;C:\Program Files\gs\gs9.52\bin;C:\Program Files\Git\cmd;C:\Users\rocha\AppData\Local\Microsoft\WindowsApps;C:\Users\rocha\AppData\Local\Programs\MiKTeX\miktex\bin\x64\;c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.PY;.PYW
```

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>mkdir helloworld
mkdir helloworld

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>cd helloworld
cd helloworld

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>vcpkg new --application
vcpkg new --application
```

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>vcpkg add port fmt
vcpkg add port fmt
Succeeded in adding ports to vcpkg.json file.

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>type vcpkg.json 
type vcpkg.json 
{
  "dependencies": [
    "fmt"
  ]
}

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>type CMakeLists.txt 
type CMakeLists.txt 
cmake_minimum_required(VERSION 3.10)

project(HelloWorld)

find_package(fmt CONFIG REQUIRED)

add_executable(HelloWorld helloworld.cpp)

target_link_libraries(HelloWorld PRIVATE fmt::fmt)

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>type helloworld.cpp 
type helloworld.cpp 
#include <fmt/core.h>

int main()
{
    fmt::print("Hello World!\n");
    return 0;
}

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>type CMakePresets.json 
type CMakePresets.json 
{
    "version": 3,
    "configurePresets": [
        {
            "name": "default",
            "binaryDir": "${sourceDir}/build",
            "cacheVariables": {
                "CMAKE_TOOLCHAIN_FILE": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake"
            }
        }
    ]
}
```


```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>cmake --preset=default
cmake --preset=default
Preset CMake variables:

  CMAKE_TOOLCHAIN_FILE="c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg/scripts/buildsystems/vcpkg.cmake"

-- Building for: Visual Studio 17 2022
-- Running vcpkg install
Fetching registry information from https://github.com/microsoft/vcpkg (HEAD)...
A suitable version of cmake was not found (required v3.27.1) Downloading portable cmake 3.27.1...
Downloading cmake...
https://github.com/Kitware/CMake/releases/download/v3.27.1/cmake-3.27.1-windows-i386.zip->C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\downloads\cmake-3.27.1-windows-i386.zip
Downloading https://github.com/Kitware/CMake/releases/download/v3.27.1/cmake-3.27.1-windows-i386.zip
Extracting cmake...
Detecting compiler hash for triplet x64-windows...
A suitable version of powershell-core was not found (required v7.2.11) Downloading portable powershell-core 7.2.11...
Downloading powershell-core...
https://github.com/PowerShell/PowerShell/releases/download/v7.2.11/PowerShell-7.2.11-win-x86.zip->C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\downloads\PowerShell-7.2.11-win-x86.zip
Downloading https://github.com/PowerShell/PowerShell/releases/download/v7.2.11/PowerShell-7.2.11-win-x86.zip
Extracting powershell-core...
The following packages will be built and installed:
    fmt:x64-windows@10.2.1 -- C:\Users\rocha\AppData\Local\vcpkg\registries\git-trees\405156a2b01c91258bf66768ceb3ae75c1caba7f
  * vcpkg-cmake:x64-windows@2023-05-04 -- C:\Users\rocha\AppData\Local\vcpkg\registries\git-trees\88a7058fc7fa73a9c4c99cfcae9d79e2abf87a5a
  * vcpkg-cmake-config:x64-windows@2022-02-06#1 -- C:\Users\rocha\AppData\Local\vcpkg\registries\git-trees\8d54cc4f487d51b655abec5f9c9c3f86ca83311f
Additional packages (*) will be modified to complete this operation.
A suitable version of 7zip was not found (required v23.1.0) Downloading portable 7zip 23.1.0...
Downloading 7zip...
https://www.7-zip.org/a/7z2301-extra.7z->C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\downloads\7z2301-extra.7z
Downloading https://www.7-zip.org/a/7z2301-extra.7z
Extracting 7zip...
Restored 3 package(s) from C:\Users\rocha\AppData\Local\vcpkg\archives in 64.6 ms. Use --debug to see more details.
Installing 1/3 vcpkg-cmake-config:x64-windows@2022-02-06#1...
Elapsed time to handle vcpkg-cmake-config:x64-windows: 4.79 ms
vcpkg-cmake-config:x64-windows package ABI: 20b02959c1badd0b0b2b18e80846b45687839ecb3ca822a7d65fb1b8142ed722
Installing 2/3 vcpkg-cmake:x64-windows@2023-05-04...
Elapsed time to handle vcpkg-cmake:x64-windows: 4.7 ms
vcpkg-cmake:x64-windows package ABI: c62da996a7b71d8c0f633c18ce2938d82903a191cc178cbee3722793543520d5
Installing 3/3 fmt:x64-windows@10.2.1...
Elapsed time to handle fmt:x64-windows: 13.5 ms
fmt:x64-windows package ABI: c87d6137cd1d670d2a0b4bae5d358af57d6c3f0973f0f38fe4e2780a971d4746
Total install time: 23.1 ms
The package fmt provides CMake targets:

    find_package(fmt CONFIG REQUIRED)
    target_link_libraries(main PRIVATE fmt::fmt)

    # Or use the header-only version
    find_package(fmt CONFIG REQUIRED)
    target_link_libraries(main PRIVATE fmt::fmt-header-only)

-- Running vcpkg install - done
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22631.
-- The C compiler identification is MSVC 19.38.33134.0
-- The CXX compiler identification is MSVC 19.38.33134.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files/Microsoft Visual Studio/2022/Community/VC/Tools/MSVC/14.38.33130/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: C:/Users/rocha/Desktop/VCPKG-TEST/vcpkg/helloworld/build
```

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>cmake --build build
cmake --build build
MSBuild version 17.8.3+195e7f5a3 for .NET Framework

  Checking Build System
  Building Custom Rule C:/Users/rocha/Desktop/VCPKG-TEST/vcpkg/helloworld/CMakeLists.txt
  helloworld.cpp
  HelloWorld.vcxproj -> C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld\build\Debug\HelloWorld.exe
  Building Custom Rule C:/Users/rocha/Desktop/VCPKG-TEST/vcpkg/helloworld/CMakeLists.txt

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>build\Debug\HelloWorld.exe
build\Debug\HelloWorld.exe
Hello World!
```

## Create and publish packages

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>cd ..
cd ..

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>mkdir custom-overlay
mkdir custom-overlay

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg>cd custom-overlay
cd custom-overlay

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>mkdir vcpkg-sample-library
mkdir vcpkg-sample-library
```

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>type vcpkg-sample-library\vcpkg.json
type vcpkg-sample-library\vcpkg.json
{
  "name": "vcpkg-sample-library",
  "version": "1.0.0",
  "homepage": "https://github.com/microsoft/vcpkg-docs/tree/cmake-sample-lib",
  "description": "A sample C++ library designed to serve as a foundational example for a tutorial on packaging libraries with vcpkg.",
  "license": "MIT",
  "dependencies": [
    {
      "name" : "vcpkg-cmake",
      "host" : true
    },
    {
      "name" : "vcpkg-cmake-config",
      "host" : true
    },
    "fmt"
  ]
}

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>type vcpkg-sample-library\usage
type vcpkg-sample-library\usage
vcpkg-sample-library provides CMake targets:

find_package(my_sample_lib CONFIG REQUIRED)
target_link_libraries(main PRIVATE my_sample_lib::my_sample_lib)
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>type vcpkg-sample-library\portfile.cmake
type vcpkg-sample-library\portfile.cmake
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)

vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO Microsoft/vcpkg-docs
    REF "${VERSION}"
    SHA512 0  # This is a temporary value. We will modify this value in the next section.
    HEAD_REF cmake-sample-lib
)


vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
)

vcpkg_cmake_install()

vcpkg_cmake_config_fixup(PACKAGE_NAME "my_sample_lib")

file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")

file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" COPYONLY)
```

**Deviation #1: Run the command so it fails and gives the _Actual hash_. The hash value does not match the example documentation**

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>vcpkg install vcpkg-sample-library --overlay-ports="%CD%"
vcpkg install vcpkg-sample-library --overlay-ports="%CD%"
warning: In the September 2023 release, the default triplet for vcpkg libraries changed from x86-windows to the detected host triplet (x64-windows). For the old behavior, add --triplet x86-windows . To suppress this message, add --triplet x64-windows .
Computing installation plan...
The following packages will be built and installed:
  * fmt:x64-windows@10.2.1
  * vcpkg-cmake:x64-windows@2023-05-04
  * vcpkg-cmake-config:x64-windows@2022-02-06#1
    vcpkg-sample-library:x64-windows@1.0.0 -- c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay\vcpkg-sample-library
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
Restored 3 package(s) from C:\Users\rocha\AppData\Local\vcpkg\archives in 60.9 ms. Use --debug to see more details.
Installing 1/4 vcpkg-cmake:x64-windows@2023-05-04...
Elapsed time to handle vcpkg-cmake:x64-windows: 4.9 ms
vcpkg-cmake:x64-windows package ABI: c62da996a7b71d8c0f633c18ce2938d82903a191cc178cbee3722793543520d5
Installing 2/4 vcpkg-cmake-config:x64-windows@2022-02-06#1...
Elapsed time to handle vcpkg-cmake-config:x64-windows: 4.24 ms
vcpkg-cmake-config:x64-windows package ABI: 20b02959c1badd0b0b2b18e80846b45687839ecb3ca822a7d65fb1b8142ed722
Installing 3/4 fmt:x64-windows@10.2.1...
Elapsed time to handle fmt:x64-windows: 13.3 ms
fmt:x64-windows package ABI: c87d6137cd1d670d2a0b4bae5d358af57d6c3f0973f0f38fe4e2780a971d4746
Installing 4/4 vcpkg-sample-library:x64-windows@1.0.0...
Building vcpkg-sample-library:x64-windows@1.0.0...
-- Installing port from location: c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay\vcpkg-sample-library
-- Note: vcpkg-sample-library only supports static library linkage. Building static library.
-- Downloading https://github.com/Microsoft/vcpkg-docs/archive/1.0.0.tar.gz -> Microsoft-vcpkg-docs-1.0.0.tar.gz...
[DEBUG] To include the environment variables in debug output, pass --debug-env
[DEBUG] Trying to load bundleconfig from c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\vcpkg-bundle.json
[DEBUG] Failed to open: c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\vcpkg-bundle.json
[DEBUG] Bundle config: readonly=false, usegitregistry=false, embeddedsha=nullopt, deployment=Git, vsversion=nullopt
[DEBUG] Metrics enabled.
[DEBUG] Feature flag 'binarycaching' unset
[DEBUG] Feature flag 'compilertracking' unset
[DEBUG] Feature flag 'registries' unset
[DEBUG] Feature flag 'versions' unset
[DEBUG] Feature flag 'dependencygraph' unset
Downloading https://github.com/Microsoft/vcpkg-docs/archive/1.0.0.tar.gz
[DEBUG] Trying to hash C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\downloads\Microsoft-vcpkg-docs-1.0.0.tar.gz.8732.part
[DEBUG] C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\downloads\Microsoft-vcpkg-docs-1.0.0.tar.gz.8732.part has hash 3f206cc2fe61d9c97c82b30852e1e4e6df299d93f6159edd1e56c644fa03ccc4670f7681e356d0e3db898a74e099a1ec531821df5430a7b14d61c743c5aa8c30
error: Failed to download from mirror set
error: File does not have the expected hash:
url: https://github.com/Microsoft/vcpkg-docs/archive/1.0.0.tar.gz
File: C:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\downloads\Microsoft-vcpkg-docs-1.0.0.tar.gz.8732.part
Expected hash: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
Actual hash: 3f206cc2fe61d9c97c82b30852e1e4e6df299d93f6159edd1e56c644fa03ccc4670f7681e356d0e3db898a74e099a1ec531821df5430a7b14d61c743c5aa8c30
[DEBUG] D:\a\_work\1\s\src\vcpkg\base\downloads.cpp(1029): 
[DEBUG] Time in subprocesses: 0us
[DEBUG] Time in parsing JSON: 5us
[DEBUG] Time in JSON reader: 0us
[DEBUG] Time in filesystem: 1027us
[DEBUG] Time in loading ports: 0us
[DEBUG] Exiting after 588 ms (578117us)

CMake Error at scripts/cmake/vcpkg_download_distfile.cmake:32 (message):
      
      Failed to download file with error: 1
      If you are using a proxy, please check your proxy setting. Possible causes are:
      
      1. You are actually using an HTTP proxy, but setting HTTPS_PROXY variable
         to `https://address:port`. This is not correct, because `https://` prefix
         claims the proxy is an HTTPS proxy, while your proxy (v2ray, shadowsocksr
         , etc..) is an HTTP proxy. Try setting `http://address:port` to both
         HTTP_PROXY and HTTPS_PROXY instead.
      
      2. If you are using Windows, vcpkg will automatically use your Windows IE Proxy Settings
         set by your proxy software. See https://github.com/microsoft/vcpkg-tool/pull/77
         The value set by your proxy might be wrong, or have same `https://` prefix issue.
      
      3. Your proxy's remote server is out of service.
      
      If you've tried directly download the link, and believe this is not a temporary
      download server failure, please submit an issue at https://github.com/Microsoft/vcpkg/issues
      to report this upstream download server failure.
      

Call Stack (most recent call first):
  scripts/cmake/vcpkg_download_distfile.cmake:270 (z_vcpkg_download_distfile_show_proxy_and_fail)
  scripts/cmake/vcpkg_from_github.cmake:106 (vcpkg_download_distfile)
  custom-overlay/vcpkg-sample-library/portfile.cmake:3 (vcpkg_from_github)
  scripts/ports.cmake:170 (include)


error: building vcpkg-sample-library:x64-windows failed with: BUILD_FAILED
Elapsed time to handle vcpkg-sample-library:x64-windows: 651 ms
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+vcpkg-sample-library
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[vcpkg-sample-library]+Build+error+on+x64-windows&body=Copy+issue+body+from+C%3A%2FUsers%2Frocha%2FDesktop%2FVCPKG-TEST%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md
```

**Update the SHA (Hash)**
```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>type vcpkg-sample-library\portfile.cmake
type vcpkg-sample-library\portfile.cmake
vcpkg_check_linkage(ONLY_STATIC_LIBRARY)

vcpkg_from_github(
    OUT_SOURCE_PATH SOURCE_PATH
    REPO Microsoft/vcpkg-docs
    REF "${VERSION}"
    SHA512 3f206cc2fe61d9c97c82b30852e1e4e6df299d93f6159edd1e56c644fa03ccc4670f7681e356d0e3db898a74e099a1ec531821df5430a7b14d61c743c5aa8c30
    HEAD_REF cmake-sample-lib
)


vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
)

vcpkg_cmake_install()

vcpkg_cmake_config_fixup(PACKAGE_NAME "my_sample_lib")

file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")

file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" COPYONLY)
```

**Deviation #2: running the install command again doesn't cause the code to compile. The example output shows that the code compiles, but my execution does not.**
```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>vcpkg install vcpkg-sample-library --overlay-ports="%CD%"
vcpkg install vcpkg-sample-library --overlay-ports="%CD%"
warning: In the September 2023 release, the default triplet for vcpkg libraries changed from x86-windows to the detected host triplet (x64-windows). For the old behavior, add --triplet x86-windows . To suppress this message, add --triplet x64-windows .
Computing installation plan...
The following packages will be built and installed:
    vcpkg-sample-library:x64-windows@1.0.0 -- c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay\vcpkg-sample-library
Detecting compiler hash for triplet x64-windows...
Restored 1 package(s) from C:\Users\rocha\AppData\Local\vcpkg\archives in 22.3 ms. Use --debug to see more details.
Installing 1/1 vcpkg-sample-library:x64-windows@1.0.0...
Elapsed time to handle vcpkg-sample-library:x64-windows: 7.49 ms
vcpkg-sample-library:x64-windows package ABI: 9a45655de051c185fd1e6c3b77b815ae7fb674354c421ef83d72c1d1f04adee3
Total install time: 7.51 ms
vcpkg-sample-library provides CMake targets:

find_package(my_sample_lib CONFIG REQUIRED)
target_link_libraries(main PRIVATE my_sample_lib::my_sample_lib)
```

```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\custom-overlay>cd ..\helloworld
cd ..\helloworld

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>type CMakeLists.txt
type CMakeLists.txt
cmake_minimum_required(VERSION 3.10)

project(HelloWorld)

find_package(fmt CONFIG REQUIRED)
find_package(my_sample_lib CONFIG REQUIRED)  # Add this line

add_executable(HelloWorld helloworld.cpp)

target_link_libraries(HelloWorld PRIVATE fmt::fmt)
target_link_libraries(HelloWorld PRIVATE my_sample_lib::my_sample_lib)  # Add this line

c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>type helloworld.cpp
type helloworld.cpp
#include "my_sample_lib.h"  // Replace #include <fmt/core.h> with "my_sample_lib.h"

int main()
{
    greet("World!");  // Replace fmt::print("Hello World!\n) with this line
    return 0;
}
```

**Deviation #3: Usage with the Hello World program fails**
```
c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg\helloworld>cmake --preset=default
cmake --preset=default
Preset CMake variables:

  CMAKE_TOOLCHAIN_FILE="c:\Users\rocha\Desktop\VCPKG-TEST\vcpkg/scripts/buildsystems/vcpkg.cmake"

-- Running vcpkg install
Detecting compiler hash for triplet x64-windows...
All requested packages are currently installed.
Total install time: 1.1 us
The package fmt provides CMake targets:

    find_package(fmt CONFIG REQUIRED)
    target_link_libraries(main PRIVATE fmt::fmt)

    # Or use the header-only version
    find_package(fmt CONFIG REQUIRED)
    target_link_libraries(main PRIVATE fmt::fmt-header-only)

-- Running vcpkg install - done
-- Selecting Windows SDK version 10.0.22000.0 to target Windows 10.0.22631.
CMake Error at C:/Users/rocha/Desktop/VCPKG-TEST/vcpkg/scripts/buildsystems/vcpkg.cmake:859 (_find_package):
  Could not find a package configuration file provided by "my_sample_lib"
  with any of the following names:

    my_sample_libConfig.cmake
    my_sample_lib-config.cmake

  Add the installation prefix of "my_sample_lib" to CMAKE_PREFIX_PATH or set
  "my_sample_lib_DIR" to a directory containing one of the above files.  If
  "my_sample_lib" provides a separate development package or SDK, be sure it
  has been installed.
Call Stack (most recent call first):
  CMakeLists.txt:6 (find_package)


-- Configuring incomplete, errors occurred!
See also "C:/Users/rocha/Desktop/VCPKG-TEST/vcpkg/helloworld/build/CMakeFiles/CMakeOutput.log".
```

--- a/vcpkg/get_started/get-started-packaging.md
+++ b/vcpkg/get_started/get-started-packaging.md
@@ -288,7 +288,7 @@ Modify `helloworld/vcpkg.json` to add a dependency on `vcpkg-sample-library`:
 ```
 
 Modify `helloworld/vcpkg-configuration.json` to include the `overlay-ports`
-folderthat contains the new port:
+folder that contains the new port:
 
 ```json
 {

--- a/vcpkg/get_started/get-started-packaging.md
+++ b/vcpkg/get_started/get-started-packaging.md
@@ -1,5 +1,5 @@
 ---
-title: "Tutorial: Packaging a library"
+title: "Tutorial: Package a library with vcpkg"
 description: Tutorial guides the user through the process of packaging a library for vcpkg.
 zone_pivot_group_filename: zone-pivot-groups.json
 zone_pivot_groups: shell-selections
@@ -10,9 +10,11 @@ ms.date: 01/10/2024
 #CustomerIntent: As a beginner C++ developer, I want to learn how to package libraries for vcpkg using custom overlays.
 ---
 
-# Tutorial: Create and publish packages
+# Tutorial: Package a library with vcpkg
 
-This tutorial guides you on how to package a library for vcpkg using a custom overlay. We recommended that you read the [Install and use packages with CMake](get-started.md) tutorial before proceeding.
+This tutorial guides you on how to package a library for vcpkg using a custom
+overlay. We recommended that you read the [Install and use packages with
+CMake](get-started.md) tutorial before proceeding.
 
 ## Prerequisites
 
@@ -24,7 +26,7 @@ This tutorial guides you on how to package a library for vcpkg using a custom ov
 ::: zone pivot="shell-cmd, shell-powershell"
 
 > [!NOTE]
-> For Windows users, Visual Studio's MSVC is the required compiler for C++ development.
+> On Windows, this tutorial uses Visual Studio's MSVC as the compiler for C++ development.
 
 ::: zone-end
 
@@ -44,29 +46,37 @@ export PATH=$VCPKG_ROOT:$PATH
 ```
 
 > [!NOTE]
-> Setting the `VCPKG_ROOT` environment variable using the `export` command only affects the current shell session. To make this change permanent across sessions, you'll need to add the `export` command to your shell's profile script (e.g., `~/.bashrc` or `~/.zshrc`).
+> Setting the `VCPKG_ROOT` environment variable using the `export` command only
+> affects the current shell session. To make this change permanent across
+> sessions, you'll need to add the `export` command to your shell's profile
+> script (e.g., `~/.bashrc` or `~/.zshrc`).
 
 ::: zone-end
 ::: zone pivot="shell-cmd"
 
 ```console
-set VCPKG_ROOT="C:\path\to\vcpkg"
+set "VCPKG_ROOT=C:\path\to\vcpkg"
 set PATH=%VCPKG_ROOT%;%PATH%
 ```
 
 > [!NOTE]
-> Setting the `VCPKG_ROOT` environment variable using the `set` command only affects the current shell session. To make this change permanent across sessions, you can use the `setx` command and restart the shell session.
+> Setting the `VCPKG_ROOT` environment variable using the `set` command only
+> affects the current shell session. To make this change permanent across
+> sessions, you can use the `setx` command and restart the shell session.
 
 ::: zone-end
 ::: zone pivot="shell-powershell"
 
 ```powershell
-$env:VCPKG_ROOT = "C:\path\to\vcpkg"
-$env:PATH = "$env:VCPKG_ROOT;$env:PATH"
+$env:VCPKG_ROOT="C:\path\to\vcpkg"
+$env:PATH="$env:VCPKG_ROOT;$env:PATH"
 ```
 
 > [!NOTE]
-> Setting the `VCPKG_ROOT` and updating the `PATH` environment variables in this manner only affects the current PowerShell session. To make these changes permanent across all sessions, you should add them to your PowerShell profile or set them through the Windows System Environment Variables panel.
+> Setting the `VCPKG_ROOT` and updating the `PATH` environment variables in this
+> manner only affects the current PowerShell session. To make these changes
+> permanent across all sessions, you should add them to your PowerShell profile
+> or set them through the Windows System Environment Variables panel.
 ::: zone-end
 
 Setting `VCPKG_ROOT` tells vcpkg where your vcpkg instance is located.
@@ -74,17 +84,20 @@ Adding it to `PATH` ensures you can run vcpkg commands directly from the shell.
 
 ## 3 - Set up the custom overlay
 
-1. Create a new directory called `custom-overlay` next to the `Hello World` project you created in the [Install and use packages with CMake](get-started.md) tutorial.
+1. Create a new directory called `custom-overlay` next to the `Hello World`
+   project you created in the [Install and use packages with
+   CMake](get-started.md) tutorial.
 2. Within `custom-overlay` directory, create a folder named `vcpkg-sample-library`.
 
 ## 4 - Set up the port files
 
-First, create the `vcpkg.json` file within the `custom-overlay\vcpkg-sample-library` folder with the following content:
+First, create the `vcpkg.json` file within the
+`custom-overlay\vcpkg-sample-library` folder with the following content:
 
 ```json
 {
   "name": "vcpkg-sample-library",
-  "version": "1.0.0",
+  "version": "1.0.2",
   "homepage": "https://github.com/microsoft/vcpkg-docs/tree/cmake-sample-lib",
   "description": "A sample C++ library designed to serve as a foundational example for a tutorial on packaging libraries with vcpkg.",
   "license": "MIT",
@@ -102,37 +115,54 @@ First, create the `vcpkg.json` file within the `custom-overlay\vcpkg-sample-libr
 }
 ```
 
-The `vcpkg.json` file serves as a manifest that defines metadata and dependencies for a C++ library, providing vcpkg with the necessary information to build, install, and manage the package.
+The `vcpkg.json` file serves as a manifest that defines metadata and
+dependencies for a C++ library, providing vcpkg with the necessary information
+to build, install, and manage the package.
 
 - `name`: Specifies the name of the library. This is used as the package identifier.
 - `version`: Indicates the version number of the library.
-- `homepage`: URL to the project's homepage, often its repository. Useful for those who want to know more or contribute.
-- `description`: Brief text describing what the library does. This is for documentation and users.
+- `homepage`: URL to the project's homepage, often its repository. Useful for
+  those who want to know more or contribute.
+- `description`: Brief text describing what the library does. This is for
+  documentation and users.
 - `license`: Specifies the license under which the library is distributed.
 - `dependencies`: An array containing the list of dependencies that the library needs.
-- `name`: `vcpkg-cmake`: Specifies a dependency on `vcpkg-cmake`, which provides CMake functions and macros commonly used in vcpkg ports.
-- `host`: true: Specifies that `vcpkg-cmake` is a host dependency, meaning it's required for building the package but not for using it.
-- `name`: `vcpkg-cmake-config`: Specifies a dependency on `vcpkg-cmake-config`, which assists in using CMake config scripts.
-- `fmt`: Specifies a run-time dependency on the `fmt` library. This means `fmt` is required for both building and using the package.
+- `name`: `vcpkg-cmake`: Specifies a dependency on `vcpkg-cmake`, which provides
+  CMake functions and macros commonly used in vcpkg ports.
+- `host`: true: Specifies that `vcpkg-cmake` is a host dependency, meaning it's
+  required for building the package but not for using it.
+- `name`: `vcpkg-cmake-config`: Specifies a dependency on `vcpkg-cmake-config`,
+  which assists in using CMake config scripts.
+- `fmt`: Specifies a run-time dependency on the `fmt` library. This means `fmt`
+  is required for both building and using the package.
 
-For more information on `vcpkg.json`, see the following documentation on [manifests](../reference/vcpkg-json.md).
+For more information on `vcpkg.json`, see the following documentation on
+[manifests](../reference/vcpkg-json.md).
 
-Now, create the `usage` file within the `custom-overlay\vcpkg-sample-library` directory with the following content:
+Now, create the `usage` file within the `custom-overlay\vcpkg-sample-library`
+directory with the following content:
 
-```
+```usage
 vcpkg-sample-library provides CMake targets:
 
 find_package(my_sample_lib CONFIG REQUIRED)
 target_link_libraries(main PRIVATE my_sample_lib::my_sample_lib)
 ```
 
-Providing usage documentation for ports allows users to easily adopt them in their projects. We highly encourage providing a `usage` file within the port's directory (`ports/<port name>/usage`) that describes the minimal steps necessary to integrate with a build system. To determine the correct usage instructions it is recommended to follow upstream's guidance. In the case that upstream does not provide usage information, it may be necessary to dig through their build system to find the exported targets.
+Providing usage documentation for ports allows users to easily adopt them in
+their projects. We highly encourage providing a `usage` file within the port's
+directory (`ports/<port name>/usage`) that describes the minimal steps necessary
+to integrate with a build system. To determine the correct usage instructions it
+is recommended to follow upstream's guidance. In the case that upstream does not
+provide usage information, it may be necessary to dig through their build system
+to find the exported targets.
 
 For more guidance, see [handling usage files](../maintainers/handling-usage-files.md)
 
-Finally, create the `portfile.cmake` file within the `custom-overlay\vcpkg-sample-library` directory with the following content:
+Finally, create the `portfile.cmake` file within the
+`custom-overlay\vcpkg-sample-library` directory with the following content:
 
-```
+```cmake
 vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
 
 vcpkg_from_github(
@@ -158,22 +188,32 @@ file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share
 configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" "${CURRENT_PACKAGES_DIR}/share/${PORT}/usage" COPYONLY)
 ```
 
-This `portfile` defines how to download, build, install, and package a specific C++ library from GitHub using vcpkg.
+This `portfile` defines how to download, build, install, and package a specific
+C++ library from GitHub using vcpkg.
 
-- `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)`: Specifies that only static linking is supported for this package.
-- `vcpkg_from_github`: Starts the function to download the source code from a GitHub repository.
-  - `OUT_SOURCE_PATH SOURCE_PATH`: Sets the directory where the source code will be extracted.
-  - `REPO JavierMatosD/vcpkg-sample-library`: The GitHub repository containing the source code.
+- `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)`: Specifies that only static linking
+  is supported for this package.
+- `vcpkg_from_github`: Starts the function to download the source code from a
+  GitHub repository.
+  - `OUT_SOURCE_PATH SOURCE_PATH`: Sets the directory where the source code will
+    be extracted.
+  - `REPO JavierMatosD/vcpkg-sample-library`: The GitHub repository containing
+    the source code.
   - `REF "${VERSION}"`: The version of the source code to download.
-  - `SHA512 0`: Placeholder for the SHA-512 hash of the source code for integrity verification.
+  - `SHA512 0`: Placeholder for the SHA-512 hash of the source code for
+    integrity verification.
   - `HEAD_REF main`: Specifies the default branch for the repository.
 - `vcpkg_cmake_configure`: Configures the project using CMake, setting up the build.
   - `SOURCE_PATH "${SOURCE_PATH}"`: The path to the source code downloaded earlier.
 - `vcpkg_cmake_install()`: Builds and installs the package using CMake.
-- `vcpkg_cmake_config_fixup(PACKAGE_NAME "my_sample_lib")`: Fixes the CMake package configuration files to be compatible with Vcpkg.
-- `file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")`: Deletes the include directory from the debug installation to prevent overlap.
-- `file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION ...)`: Installs the LICENSE file to the package's share directory and renames it to copyright.
-- `configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" ...)`: Copies a usage instruction file to the package's share directory.
+- `vcpkg_cmake_config_fixup(PACKAGE_NAME "my_sample_lib")`: Fixes the CMake
+  package configuration files to be compatible with vcpkg.
+- `file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")`: Deletes the
+  include directory from the debug installation to prevent overlap.
+- `file(INSTALL "${SOURCE_PATH}/LICENSE" DESTINATION ...)`: Installs the LICENSE
+  file to the package's share directory and renames it to copyright.
+- `configure_file("${CMAKE_CURRENT_LIST_DIR}/usage" ...)`: Copies a usage
+  instruction file to the package's share directory.
 
 For more information, refer to the [maintainer guide](../contributing/maintainer-guide.md).
 
@@ -189,10 +229,12 @@ You will get a long error message. Scan the output until you find:
 
 ```console
 Expected hash: 00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
-Actual hash: 56cffcf9bfb6e9c6bc41e28ae58e8ad25c112db49eafa63f20f95c1ced5712ba7bada9e0990741f70a3f45a4cfd6cb25a3500ef04d36037309b1b65e2f8b569d
+Actual hash: 4202125968a01219deeee14b81e1d476dab18d968425ba36d640816b0b3db6168f8ccf4120ba20526e9930c8c7294e64d43900ad2aef9d5f28175210d0c3a417
 ```
 
-Copy the "Actual hash"  `56cffcf9bfb6e9c6bc41e28ae58e8ad25c112db49eafa63f20f95c1ced5712ba7bada9e0990741f70a3f45a4cfd6cb25a3500ef04d36037309b1b65e2f8b569d`, and replace the `SHA512` value in the `portfile.cmake` with its value.
+Copy the "Actual hash"
+`4202125968a01219deeee14b81e1d476dab18d968425ba36d640816b0b3db6168f8ccf4120ba20526e9930c8c7294e64d43900ad2aef9d5f28175210d0c3a417`,
+and replace the `SHA512` value in the `portfile.cmake` with its value.
 
 Re-run the install command:
 
@@ -200,38 +242,41 @@ Re-run the install command:
 vcpkg install vcpkg-sample-library --overlay-ports=C:\path\to\custom-overlay
 ```
 
-```
+```console
 Computing installation plan...
 The following packages will be built and installed:
-    vcpkg-sample-library:x64-windows -> 1.0.3
+    vcpkg-sample-library:x64-windows -> 1.0.2 -- C:\Users\dev\demo\custom-overlay\vcpkg-sample-library
 Detecting compiler hash for triplet x64-windows...
-Restored 0 package(s) from C:\Users\javiermat\AppData\Local\vcpkg\archives in 132 us. Use --debug to see more details.
+Restored 0 package(s) from C:\Users\dev\AppData\Local\vcpkg\archives in 174 us. Use --debug to see more details.
 Installing 1/1 vcpkg-sample-library:x64-windows...
 Building vcpkg-sample-library:x64-windows...
+-- Installing port from location: C:\Users\dev\demo\custom-overlay\vcpkg-sample-library
 -- Note: vcpkg-sample-library only supports static library linkage. Building static library.
--- Downloading https://github.com/JavierMatosD/vcpkg-sample-library/archive/1.0.3.tar.gz -> JavierMatosD-vcpkg-sample-library-1.0.3.tar.gz...
--- Extracting source C:/dev/vcpkg/downloads/JavierMatosD-vcpkg-sample-library-1.0.3.tar.gz
--- Using source at C:/dev/vcpkg/buildtrees/vcpkg-sample-library/src/1.0.3-8d09a66448.clean
--- Found external ninja('1.11.0').
+-- Using cached Microsoft-vcpkg-docs-1.0.2.tar.gz.
+-- Cleaning sources at C:/Users/dev/demo/vcpkg/buildtrees/vcpkg-sample-library/src/1.0.2-2aff836404.clean. Use --editable to skip cleaning for the packages you specify.
+-- Extracting source C:/Users/dev/demo/vcpkg/downloads/Microsoft-vcpkg-docs-1.0.2.tar.gz
+-- Using source at C:/Users/dev/demo/vcpkg/buildtrees/vcpkg-sample-library/src/1.0.2-2aff836404.clean
 -- Configuring x64-windows
 -- Building x64-windows-dbg
 -- Building x64-windows-rel
--- Installing: C:/dev/vcpkg/packages/vcpkg-sample-library_x64-windows/share/vcpkg-sample-library/copyright
+-- Installing: C:/Users/dev/demo/vcpkg/packages/vcpkg-sample-library_x64-windows/share/vcpkg-sample-library/copyright
 -- Performing post-build validation
-Stored binaries in 1 destinations in 111 ms.
-Elapsed time to handle vcpkg-sample-library:x64-windows: 4.4 s
-Total install time: 4.4 s
+Stored binaries in 1 destinations in 94 ms.
+Elapsed time to handle vcpkg-sample-library:x64-windows: 6.1 s
+Total install time: 6.1 s
 vcpkg-sample-library provides CMake targets:
 
-  find_package(my_sample_lib CONFIG REQUIRED)
-  target_link_libraries(main PRIVATE my_sample_lib::my_sample_lib)
+find_package(my_sample_lib CONFIG REQUIRED)
+target_link_libraries(main PRIVATE my_sample_lib::my_sample_lib)
 ```
 
 ## 6 - Verify the port build
 
-To verify the library builds and links properly, add the new dependency to the manifest and configuration files.
+To verify the library builds and links properly, add a new dependency to the
+`helloworld` project created in the [install packages tutorial](get-started.md).
+Make the following changes to the project's manifest and configuration files.
 
-Modify `vcpkg.json` to add the `vcpkg-sample-library` dependency:
+Modify `helloworld/vcpkg.json` to add a dependency on `vcpkg-sample-library`:
 
 ```json
 {
@@ -242,7 +287,8 @@ Modify `vcpkg.json` to add the `vcpkg-sample-library` dependency:
 }
 ```
 
-Modify `vcpkg-configuration.json` to include the `overlay-ports` containing the new port:
+Modify `helloworld/vcpkg-configuration.json` to include the `overlay-ports`
+folderthat contains the new port:
 
 ```json
 {
@@ -259,15 +305,15 @@ Modify `vcpkg-configuration.json` to include the `overlay-ports` containing the 
     }
   ],
   "overlay-ports": [
-    "../custom-overlay/vcpkg-sample-library"
+    "../custom-overlay"
   ]
 }
 ```
 
-Next, modify the `CMakeLists.txt` and `main.cpp` created in the [Install and use packages with
-CMake](get-started.md) tutorial.
+Next, modify `helloworld/CMakeLists.txt` and `helloworld/main.cpp` to use the
+new dependency.
 
-Modify the `CMakeLists.txt` with the following content:
+Modify the `helloworld/CMakeLists.txt` with the following content:
 
 ```cmake
 cmake_minimum_required(VERSION 3.10)
@@ -290,7 +336,7 @@ Modify the `main.cpp` with the following content:
 
 int main()
 {
-    greet("World!");  // Replace fmt::print("Hello World!\n) with this line
+    greet("vcpkg!");  // Replace fmt::print("Hello World!\n) with this line
     return 0;
 }
 
@@ -316,13 +362,17 @@ Configure, build, and run the application.
   ./build/HelloWorld
   ```
 
-  ```
-  Hello World!
+  The path to your project's executable may be different, for example: `./build/Debug/HelloWorld.exe`.
+
+  ```console
+  Hello vcpkg!
   ```
 
 ## Next steps
 
-Now that the `vcpkg-sample-library` has been packaged as port, the next step is to add it to the vcpkg curated registry. See, [Adding ports to vcpkg registry](get-started-adding-to-registry.md)
+Now that the `vcpkg-sample-library` has been packaged as port, the next step is
+to add it to the vcpkg curated registry. See, [Adding ports to vcpkg
+registry](get-started-adding-to-registry.md).
 
 For more information, see:
 

